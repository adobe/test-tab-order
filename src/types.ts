import * as puppeteer from 'puppeteer'

export interface ConfigureOptions {
  maxRetriesCount: number
  delay: number
  timeout: number
}

export interface TestTabOrderOptions {
  elements?: string[]
  page: puppeteer.Page
  frame?: puppeteer.Frame
  startElement?: string
  maxTabStopsCount?: number
  delay?: number
}

interface CommonTabStopsOptions {
  tabFunction: () => Promise<void>
  frame: puppeteer.Frame | puppeteer.Page
  delay: number
}

export interface GetTabStopsOptions extends CommonTabStopsOptions {
  maxTabStopsCount: number
}

export interface CheckTabStopsOptions extends CommonTabStopsOptions {
  elements: string[]
  maxRetriesCount: number
}
