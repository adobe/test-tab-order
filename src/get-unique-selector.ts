/* eslint-disable */

/**
 * This function has all the code directly pasted from @cypress/unique-selector and its dependency, css.escape.
 */
export function getUniqueSelector(): string {
  const selectors = 'data-testid data-id id class tag attributes nth-child'.split(
    ' ',
  )
  // all imports
  var cssEscape = function (value: any) {
    if (arguments.length == 0) {
      throw new TypeError('`cssEscape` requires an argument.')
    }
    var string = String(value)
    var length = string.length
    var index = -1
    var codeUnit
    var result = ''
    var firstCodeUnit = string.charCodeAt(0)
    while (++index < length) {
      codeUnit = string.charCodeAt(index)
      // Note: there’s no need to special-case astral symbols, surrogate
      // pairs, or lone surrogates.

      // If the character is NULL (U+0000), then the REPLACEMENT CHARACTER
      // (U+FFFD).
      if (codeUnit == 0x0000) {
        result += '\uFFFD'
        continue
      }

      if (
        // If the character is in the range [\1-\1F] (U+0001 to U+001F) or is
        // U+007F, […]
        (codeUnit >= 0x0001 && codeUnit <= 0x001f) ||
        codeUnit == 0x007f ||
        // If the character is the first character and is in the range [0-9]
        // (U+0030 to U+0039), […]
        (index == 0 && codeUnit >= 0x0030 && codeUnit <= 0x0039) ||
        // If the character is the second character and is in the range [0-9]
        // (U+0030 to U+0039) and the first character is a `-` (U+002D), […]
        (index == 1 &&
          codeUnit >= 0x0030 &&
          codeUnit <= 0x0039 &&
          firstCodeUnit == 0x002d)
      ) {
        // https://drafts.csswg.org/cssom/#escape-a-character-as-code-point
        result += '\\' + codeUnit.toString(16) + ' '
        continue
      }

      if (
        // If the character is the first character and is a `-` (U+002D), and
        // there is no second character, […]
        index == 0 &&
        length == 1 &&
        codeUnit == 0x002d
      ) {
        result += '\\' + string.charAt(index)
        continue
      }

      // If the character is not handled by one of the above rules and is
      // greater than or equal to U+0080, is `-` (U+002D) or `_` (U+005F), or
      // is in one of the ranges [0-9] (U+0030 to U+0039), [A-Z] (U+0041 to
      // U+005A), or [a-z] (U+0061 to U+007A), […]
      if (
        codeUnit >= 0x0080 ||
        codeUnit == 0x002d ||
        codeUnit == 0x005f ||
        (codeUnit >= 0x0030 && codeUnit <= 0x0039) ||
        (codeUnit >= 0x0041 && codeUnit <= 0x005a) ||
        (codeUnit >= 0x0061 && codeUnit <= 0x007a)
      ) {
        // the character itself
        result += string.charAt(index)
        continue
      }

      // Otherwise, the escaped character.
      // https://drafts.csswg.org/cssom/#escape-a-character
      result += '\\' + string.charAt(index)
    }
    return result
  }
  function getAttributes(
    el: any,
    attributesToIgnore = ['id', 'class', 'length'],
  ) {
    const {attributes} = el
    const attrs = [...attributes]

    return attrs.reduce((sum, next) => {
      if (!(attributesToIgnore.indexOf(next.nodeName) > -1)) {
        sum.push(`[${next.nodeName}="${next.value}"]`)
      }
      return sum
    }, [])
  }
  function getClasses(el: any) {
    if (!el.hasAttribute('class')) {
      return []
    }

    try {
      return Array.prototype.slice.call(el.classList)
    } catch (e) {
      let className = el.getAttribute('class')

      // remove duplicate and leading/trailing whitespaces
      className = className.trim().replace(/\s+/g, ' ')

      // split into separate classnames
      return className.split(' ')
    }
  }

  /**
   * Returns the Class selectors of the element
   * @param  { Object } element
   * @return { Array }
   */
  function getClassSelectors(el: any) {
    const classList = getClasses(el).filter(Boolean)
    return classList.map((cl: any) => `.${cssEscape(cl)}`)
  }

  function kCombinations(
    result: any,
    items: any,
    data: any,
    start: any,
    end: any,
    index: any,
    k: any,
  ) {
    if (index === k) {
      result.push(data.slice(0, index).join(''))
      return
    }

    for (let i = start; i <= end && end - i + 1 >= k - index; ++i) {
      data[index] = items[i]
      kCombinations(result, items, data, i + 1, end, index + 1, k)
    }
  }

  /**
   * Returns all the possible selector combinations
   * @param  { Array } items
   * @param  { Number } k
   * @return { Array }
   */
  function getCombinations(items: any, k: any) {
    const result: any = [],
      n = items.length,
      data: any = []

    for (var l = 1; l <= k; ++l) {
      kCombinations(result, items, data, 0, n - 1, 0, l)
    }

    return result
  }
  function needsQuote(v: any) {
    // if the escaped value is different from
    // the non escaped value then we know
    // we need to quote the value
    return v !== cssEscape(v)
  }
  const getData = (selectorType: any, attributes: any) => {
    for (let i = 0; i < attributes.length; i++) {
      // extract node name + value
      const {nodeName, value} = attributes[i]

      // if this matches our selector
      if (nodeName === selectorType) {
        if (value) {
          if (needsQuote(value)) {
            // if we have value that needs quotes
            return `[${nodeName}="${value}"]`
          }

          return `[${nodeName}=${value}]`
        }

        return `[${nodeName}]`
      }
    }

    return null
  }
  function getID(el: any) {
    const id = el.getAttribute('id')

    if (id !== null && id !== '') {
      return `#${cssEscape(id)}`
    }
    return null
  }
  function getNthChild(element: any) {
    let counter = 0
    let k
    let sibling
    const {parentNode} = element

    if (Boolean(parentNode)) {
      const {childNodes} = parentNode
      const len = childNodes.length
      for (k = 0; k < len; k++) {
        sibling = childNodes[k]
        if (isElement(sibling)) {
          counter++
          if (sibling === element) {
            return `:nth-child(${counter})`
          }
        }
      }
    }
    return null
  }
  function getParents(el: any) {
    const parents = []
    let currentElement = el
    while (isElement(currentElement)) {
      parents.push(currentElement)
      currentElement = currentElement.parentNode
    }

    return parents
  }
  function getTag(el: any) {
    return el.tagName.toLowerCase().replace(/:/g, '\\:')
  }
  function isUnique(el: any, selector: any) {
    if (!Boolean(selector)) return false
    const elems = el.ownerDocument.querySelectorAll(selector)
    return elems.length === 1 && elems[0] === el
  }
  function isElement(el: any) {
    let isElem

    if (typeof HTMLElement === 'object') {
      isElem = el instanceof HTMLElement
    } else {
      isElem =
        !!el &&
        typeof el === 'object' &&
        el.nodeType === 1 &&
        typeof el.nodeName === 'string'
    }
    return isElem
  }

  const dataRegex = /^data-(.+)/

  /**
   * Returns all the selectors of the elmenet
   * @param  { Object } element
   * @return { Object }
   */
  function getAllSelectors(el: any, selectors: any, attributesToIgnore: any) {
    const funcs = {
      tag: getTag,
      'nth-child': getNthChild,
      attributes: (elem: any) => getAttributes(elem, attributesToIgnore),
      class: getClassSelectors,
      id: getID,
    } as any

    return selectors
      .filter((selector: any) => !dataRegex.test(selector))
      .reduce((res: any, next: any) => {
        res[next] = funcs[next](el)
        return res
      }, {})
  }

  /**
   * Tests uniqueNess of the element inside its parent
   * @param  { Object } element
   * @param { String } Selectors
   * @return { Boolean }
   */
  function testUniqueness(element: any, selector: any) {
    const {parentNode} = element
    const elements = parentNode.querySelectorAll(selector)
    return elements.length === 1 && elements[0] === element
  }

  /**
   * Tests all selectors for uniqueness and returns the first unique selector.
   * @param  { Object } element
   * @param  { Array } selectors
   * @return { String }
   */
  function getFirstUnique(element: any, selectors: any) {
    return selectors.find(testUniqueness.bind(null, element))
  }

  /**
   * Checks all the possible selectors of an element to find one unique and return it
   * @param  { Object } element
   * @param  { Array } items
   * @param  { String } tag
   * @return { String }
   */
  function getUniqueCombination(element: any, items: any, tag: any) {
    let combinations = getCombinations(items, 3),
      firstUnique = getFirstUnique(element, combinations)

    if (Boolean(firstUnique)) {
      return firstUnique
    }

    if (Boolean(tag)) {
      combinations = combinations.map((combination: any) => tag + combination)
      firstUnique = getFirstUnique(element, combinations)

      if (Boolean(firstUnique)) {
        return firstUnique
      }
    }

    return null
  }

  /**
   * Returns a uniqueSelector based on the passed options
   * @param  { DOM } element
   * @param  { Array } options
   * @return { String }
   */
  function getUniqueSelector(
    element: any,
    selectorTypes: any,
    attributesToIgnore: any,
  ) {
    let foundSelector

    const attributes = [...element.attributes]

    const elementSelectors = getAllSelectors(
      element,
      selectorTypes,
      attributesToIgnore,
    )

    for (let selectorType of selectorTypes) {
      let selector = elementSelectors[selectorType]

      // if we are a data attribute
      if (dataRegex.test(selectorType)) {
        const dataSelector = getData(selectorType, attributes)

        // if we found a selector via data attributes
        if (dataSelector) {
          selector = dataSelector
          selectorType = 'data'
        }
      }

      if (!Boolean(selector)) continue

      switch (selectorType) {
        case 'data':
        case 'id':
        case 'tag':
          if (testUniqueness(element, selector)) {
            return selector
          }
          break
        case 'class':
        case 'attributes':
          if (selector.length) {
            foundSelector = getUniqueCombination(
              element,
              selector,
              elementSelectors.tag,
            )
            if (foundSelector) {
              return foundSelector
            }
          }
          break

        case 'nth-child':
          return selector

        default:
          break
      }
    }
    return '*'
  }

  /**
   * Generate unique CSS selector for given DOM element
   *
   * @param {Element} el
   * @return {String}
   * @api private
   */

  function unique(el: any, options = {}): string {
    const {
      selectorTypes = ['id', 'class', 'tag', 'nth-child'],
      attributesToIgnore = ['id', 'class', 'length'],
    }: any = options
    const allSelectors = []
    const parents = getParents(el)

    for (let elem of parents) {
      const selector = getUniqueSelector(
        elem,
        selectorTypes,
        attributesToIgnore,
      )
      if (Boolean(selector)) {
        allSelectors.push(selector)
      }
    }

    const selectors = []
    for (let it of allSelectors) {
      selectors.unshift(it)
      const selector = selectors.join(' > ')
      if (isUnique(el, selector)) {
        return selector
      }
    }

    return ''
  }

  return unique(document.activeElement, {selectorTypes: selectors})
}
