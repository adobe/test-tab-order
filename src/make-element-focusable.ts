export default function makeElementFocusable(element: Element): boolean {
  function checkElementIfFocusable(element: Element) {
    const elementTagName = element.tagName
    if (
      ['A', 'LINK'].indexOf(elementTagName) > -1 &&
      element.hasAttribute('href')
    ) {
      return true
    }

    if (
      elementTagName === 'INPUT' &&
      element.getAttribute('type') !== 'hidden'
    ) {
      return true
    }

    return (
      ['BUTTON', 'SELECT', 'TEXTAREA'].indexOf(elementTagName) > -1 ||
      element.hasAttribute('tabindex')
    )
  }

  const isElementFocusable = checkElementIfFocusable(element)

  if (!isElementFocusable) {
    element.setAttribute('tabindex', '-1')
  }

  ;(element as HTMLElement).focus()

  return isElementFocusable
}
