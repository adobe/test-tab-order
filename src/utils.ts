import {ElementHandle, Frame, Page} from 'puppeteer'

export async function getActiveElement(
  frame: Frame | Page,
): Promise<ElementHandle> {
  const element = await frame.evaluateHandle<ElementHandle>(
    () => document.activeElement,
  )

  return element
}

export function getTabFunction(
  page: Page,
  backwards = false,
): () => Promise<void> {
  if (backwards) {
    return async function () {
      await page.keyboard.down('Shift')
      await page.keyboard.press('Tab')
      await page.keyboard.up('Shift')
    }
  } else {
    return async function () {
      await page.keyboard.press('Tab')
    }
  }
}
