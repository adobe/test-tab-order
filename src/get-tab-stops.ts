import {getUniqueSelector} from './get-unique-selector'
import {GetTabStopsOptions} from './types'

export default async function getTabStops(
  options: GetTabStopsOptions,
): Promise<string[]> {
  const tabStopsSlectors: string[] = []
  const {tabFunction, frame, delay, maxTabStopsCount} = options

  for (let index = 0; index < maxTabStopsCount; index++) {
    await tabFunction()

    const elementSelector = await frame.evaluate(getUniqueSelector)

    if (elementSelector === 'body') {
      return tabStopsSlectors
    } else {
      tabStopsSlectors.push(elementSelector)
    }

    await frame.waitForTimeout(delay)
  }

  if ((await frame.evaluate(getUniqueSelector)) !== 'body') {
    throw new Error(
      `We are generating selectors but we did not reach the end of the page. Consider increasing maxTabStopsCount currently at ${maxTabStopsCount}`,
    )
  }

  return tabStopsSlectors
}
