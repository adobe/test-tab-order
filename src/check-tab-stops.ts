import {getUniqueSelector} from './get-unique-selector'
import {CheckTabStopsOptions} from './types'
import {getActiveElement} from './utils'

export default async function checkTabStops(
  options: CheckTabStopsOptions,
): Promise<void> {
  const {tabFunction, elements, frame, delay, maxRetriesCount} = options
  const tabStopsCount = elements.length
  let previousActiveElementHandle = await getActiveElement(frame)
  let retriesCount = 0

  for (let index = 0; index < tabStopsCount; index++) {
    const expectedSelector = elements[index]
    await frame.waitForSelector(expectedSelector, {visible: true})

    await tabFunction()

    const actualActiveElement = await getActiveElement(frame)
    const expectedActiveElement = await frame.$(expectedSelector)

    const isSameElement = await frame.evaluate(
      (e1, e2) => e1 === e2,
      actualActiveElement,
      expectedActiveElement,
    )

    if (!isSameElement) {
      const lostFocusToBody = await frame.evaluate(() => {
        if (document.activeElement === document.body) {
          return true
        }
        return false
      })

      if (lostFocusToBody) {
        if (retriesCount++ < maxRetriesCount) {
          index-- // retry iteration

          console.warn(
            `Lost focus when checking ${expectedSelector}. Retrying.`,
          )
          await previousActiveElementHandle.focus()
        } else {
          throw new Error(
            `Focused was lost for some reason and the maxRetriesCount was exceeded.`,
          )
        }
      } else {
        throw new Error(
          `Expected element "${expectedSelector}" but found "${await frame.evaluate(
            getUniqueSelector,
          )}"`,
        )
      }
    }

    previousActiveElementHandle = actualActiveElement
    await frame.waitForTimeout(delay)
  }
}
