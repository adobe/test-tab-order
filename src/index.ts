import makeElementFocusable from './make-element-focusable'
import checkTabStops from './check-tab-stops'
import getTabStops from './get-tab-stops'
import {getTabFunction} from './utils'
import {ConfigureOptions, TestTabOrderOptions} from './types'

let configOptions: ConfigureOptions = {
  maxRetriesCount: 1,
  delay: 0,
  timeout: 3000,
}

export function configure(options: Partial<ConfigureOptions>): void {
  configOptions = {...configOptions, ...options}
}

export async function testTabOrder(
  options: TestTabOrderOptions,
): Promise<void> {
  const {
    page,
    elements,
    frame = page,
    startElement = frame ? 'body' : null,
    delay = configOptions.delay,
    maxTabStopsCount = 100,
  } = options
  let isStartElementFocusable

  if (startElement) {
    await frame.waitForSelector(startElement, {
      visible: true,
      timeout: configOptions.timeout,
    })
    isStartElementFocusable = await frame.$eval(
      startElement,
      makeElementFocusable,
    )
  }

  if (elements === undefined) {
    console.log(
      'The "elements" prop was not passed, so we went ahead and grabbed them for you:\n',
      await getTabStops({
        frame,
        tabFunction: getTabFunction(page),
        delay,
        maxTabStopsCount,
      }),
    )
  } else {
    await checkTabStops({
      frame,
      tabFunction: getTabFunction(page),
      elements,
      delay,
      maxRetriesCount: configOptions.maxRetriesCount,
    })

    await checkTabStops({
      frame,
      tabFunction: getTabFunction(page, true),
      elements: [...elements].reverse().slice(1),
      delay,
      maxRetriesCount: configOptions.maxRetriesCount,
    })
  }

  if (startElement && !isStartElementFocusable) {
    await frame.$eval(startElement, (el: Element) => {
      el.removeAttribute('tabindex')
    })
  }
}
