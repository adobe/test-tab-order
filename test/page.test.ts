import * as path from 'path'
import * as puppeteer from 'puppeteer'

import {configure, testTabOrder} from '../src'
import {elements} from './test-utils'

let browser: puppeteer.Browser
let page: puppeteer.Page

jest.setTimeout(60000)

beforeAll(async () => {
  browser = await puppeteer.launch({
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
    ],
  })
  page = await browser.newPage()
  await page.goto(`file://${path.join(__dirname, 'fixtures/page.html')}`)
})

afterAll(async () => {
  await browser.close()
})

afterEach(jest.restoreAllMocks)

test('should print to console the elements if they are not passed', async () => {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  jest.spyOn(console, 'log').mockImplementation(() => {})
  await testTabOrder({page})

  expect(console.log).toHaveBeenCalledTimes(1)
  expect((console.log as jest.Mock).mock.calls[0]).toMatchSnapshot()
})

test('should show an error if elements are not passed and maxTabStops is too small', async () => {
  await expect(
    testTabOrder({
      page,
      maxTabStopsCount: 2,
    }),
  ).rejects.toThrowErrorMatchingSnapshot()
})

test('should check the tab order is correct', async () => {
  await expect(
    testTabOrder({
      page,
      elements,
    }),
  ).resolves.not.toThrow()
})

test('should throw an error if tab order is not correct', async () => {
  await expect(
    testTabOrder({
      page,
      elements: [elements[0], elements[2]],
    }),
  ).rejects.toThrowErrorMatchingSnapshot()
})

test('should start tab order testing from a specific element', async () => {
  await expect(
    testTabOrder({
      page,
      startElement: elements[2],
      elements: elements.slice(3),
    }),
  ).resolves.not.toThrow()
})

test('should wait for element to become visible before testing', async () => {
  await page.evaluate(elementToHide => {
    document.querySelector<HTMLElement>(elementToHide).style.display = 'none'
  }, elements[0])
  const testTabOrderPromise = testTabOrder({
    page,
    elements,
  })
  await page.evaluate(elementToHide => {
    document.querySelector<HTMLElement>(elementToHide).style.display = 'inline'
  }, elements[0])

  await expect(testTabOrderPromise).resolves.not.toThrow()
})

test('delay parameter should be used in waitForTimeout', async () => {
  const delay = 600
  const waitForTimeoutSpy = jest
    .spyOn(page, 'waitForTimeout')
    .mockResolvedValue()
  const waitForTimeoutTimesCalled = elements.length * 2 - 1

  await testTabOrder({
    page,
    elements,
    delay,
  })

  expect(waitForTimeoutSpy).toHaveBeenCalledTimes(waitForTimeoutTimesCalled)
  for (let index = 1; index <= waitForTimeoutTimesCalled; index++) {
    expect(waitForTimeoutSpy).toHaveBeenNthCalledWith(index, delay)
  }
})

test('should throw error if focus is lost more than maxRetriesCount allows', async () => {
  const consoleWarnSpy = jest
    .spyOn(console, 'warn')
    .mockImplementationOnce(() => {
      // nothing
    })
  configure({maxRetriesCount: 0})

  jest.spyOn(page.keyboard, 'press').mockImplementation(async () => {
    document.body.focus()

    return Promise.resolve()
  })

  await expect(
    testTabOrder({
      page,
      elements,
    }),
  ).rejects.toThrowErrorMatchingSnapshot()
  expect(consoleWarnSpy).not.toHaveBeenCalled()
})

test('should show warning if focus is lost less than maxRetriesCount allows', async () => {
  const consoleWarnSpy = jest
    .spyOn(console, 'warn')
    .mockImplementationOnce(() => {
      // nothing
    })
  configure({maxRetriesCount: 1})

  jest.spyOn(page.keyboard, 'press').mockImplementationOnce(async () => {
    document.body.focus()

    return Promise.resolve()
  })

  await expect(
    testTabOrder({
      page,
      elements,
    }),
  ).resolves.not.toThrow()
  expect(consoleWarnSpy).toHaveBeenCalledTimes(1)
  expect(consoleWarnSpy.mock.calls[0]).toMatchSnapshot()
})
