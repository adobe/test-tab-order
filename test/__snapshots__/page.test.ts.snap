// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print to console the elements if they are not passed 1`] = `
Array [
  "The \\"elements\\" prop was not passed, so we went ahead and grabbed them for you:
",
  Array [
    "[data-testid=click-me-button]",
    "[data-testid=ok-button]",
    "[data-testid=this-page-link]",
    "[data-testid=username-input]",
    "[data-testid=password-input]",
    "[data-testid=submit-button]",
    "[data-testid=yes-no-select]",
    "[data-testid=button-as-span]",
    "[data-testid=comments-textarea]",
  ],
]
`;

exports[`should show an error if elements are not passed and maxTabStops is too small 1`] = `"We are generating selectors but we did not reach the end of the page. Consider increasing maxTabStopsCount currently at 2"`;

exports[`should show warning if focus is lost less than maxRetriesCount allows 1`] = `
Array [
  "Lost focus when checking [data-testid=click-me-button]. Retrying.",
]
`;

exports[`should throw an error if tab order is not correct 1`] = `"Expected element \\"[data-testid=this-page-link]\\" but found \\"[data-testid=ok-button]\\""`;

exports[`should throw error if focus is lost more than maxRetriesCount allows 1`] = `"Focused was lost for some reason and the maxRetriesCount was exceeded."`;
